x/auth/vesting/internal/types/vesting_account.go

```diff
@@ -0,0 +1,720 @@
+		EndTime:          endTime,
+	}
+}
+
+// String implements fmt.Stringer
+func (bva BaseVestingAccount) String() string {
+	var pubkey string
+
+	if bva.PubKey != nil {
+		pubkey = sdk.MustBech32ifyAccPub(bva.PubKey)
+	}
+
+	return fmt.Sprintf(`Vesting Account:
+  Address:          %s
+  Pubkey:           %s
+  Coins:            %s
+  AccountNumber:    %d
+  Sequence:         %d
+  OriginalVesting:  %s
+  DelegatedFree:    %s
+  DelegatedVesting: %s
+  EndTime:          %d `,
+		bva.Address, pubkey, bva.Coins, bva.AccountNumber, bva.Sequence,
+		bva.OriginalVesting, bva.DelegatedFree, bva.DelegatedVesting, bva.EndTime,
+	)
+}
+
+// spendableCoins returns all the spendable coins for a vesting account given a
+// set of vesting coins.
+//
+// CONTRACT: The account's coins, delegated vesting coins, vestingCoins must be
+// sorted.
+func (bva BaseVestingAccount) spendableCoins(vestingCoins sdk.Coins) sdk.Coins {
+	var spendableCoins sdk.Coins
+	bc := bva.GetCoins()
+
+	for _, coin := range bc {
+		// zip/lineup all coins by their denomination to provide O(n) time
+		baseAmt := coin.Amount
+		vestingAmt := vestingCoins.AmountOf(coin.Denom)
+		delVestingAmt := bva.DelegatedVesting.AmountOf(coin.Denom)
+
+		// compute min((BC + DV) - V, BC) per the specification
+		min := sdk.MinInt(baseAmt.Add(delVestingAmt).Sub(vestingAmt), baseAmt)
+		spendableCoin := sdk.NewCoin(coin.Denom, min)
+
+		if !spendableCoin.IsZero() {
+			spendableCoins = spendableCoins.Add(sdk.Coins{spendableCoin})
+		}
+	}
+
+	return spendableCoins
+}
+
+// trackDelegation tracks a delegation amount for any given vesting account type
+// given the amount of coins currently vesting. It returns the resulting base
+// coins.
+//
+// CONTRACT: The account's coins, delegation coins, vesting coins, and delegated
+// vesting coins must be sorted.
+func (bva *BaseVestingAccount) trackDelegation(vestingCoins, amount sdk.Coins) {
+	bc := bva.GetCoins()
+
+	for _, coin := range amount {
+		// zip/lineup all coins by their denomination to provide O(n) time
+
+		baseAmt := bc.AmountOf(coin.Denom)
+		vestingAmt := vestingCoins.AmountOf(coin.Denom)
+		delVestingAmt := bva.DelegatedVesting.AmountOf(coin.Denom)
+
+		// Panic if the delegation amount is zero or if the base coins does not
+		// exceed the desired delegation amount.
+		if coin.Amount.IsZero() || baseAmt.LT(coin.Amount) {
+			panic("delegation attempt with zero coins or insufficient funds")
+		}
+
+		// compute x and y per the specification, where:
+		// X := min(max(V - DV, 0), D)
+		// Y := D - X
+		x := sdk.MinInt(sdk.MaxInt(vestingAmt.Sub(delVestingAmt), sdk.ZeroInt()), coin.Amount)
+		y := coin.Amount.Sub(x)
+
+		if !x.IsZero() {
+			xCoin := sdk.NewCoin(coin.Denom, x)
+			bva.DelegatedVesting = bva.DelegatedVesting.Add(sdk.Coins{xCoin})
+		}
+
+		if !y.IsZero() {
+			yCoin := sdk.NewCoin(coin.Denom, y)
+			bva.DelegatedFree = bva.DelegatedFree.Add(sdk.Coins{yCoin})
+		}
+
+		bva.Coins = bva.Coins.Sub(sdk.Coins{coin})
```

Comment:
Although the naming/separation of concerns is a bit messy, I'm hesitant to make the `TrackDelegation` function not atomically update all the relevant balance fields on the account. If we remove the call to `Sub`, someone calling `TrackDelegation` from outside the `Bank` module could forget to subsequently call `Sub`, which would lead to the potential for coin inflation. 

For clarity, we could rename these functions to something like `DelegateAndUpdateBalance`.

What are others' thoughts?

cc @fedekunze @alexanderbez @rigelrozanski

Reply:
Largely because the `track*` prefix to me suggests that the function is intended to just change the tracking fields, notably `DelegatedVesting` and `DelegatedFree`.

I'm fine with the function doing both things, though i would prefer that the name of the function reflect this fact